### PR TITLE
colexec: fix hash and merge joiners when types of inputs in eq cols are different

### LINKED ISSUE #44797: colexec: incorrect output of hash joiner
When #44796 merges, such incorrect behavior will occur (before that PR we would just crash):
```
create table l(a int);
insert into l values (-1);
create table r (b smallint);
insert into r values (-1);

set vectorize=experimental_on;
select * from l, r where a = b;
  a | b
----+----
(0 rows)

set vectorize=off;
select * from l, r where a = b;
  a  | b
-----+-----
  -1 | -1
```

### LINKED ISSUE #44798: colexec: merge joiner incorrectly assumes the types are the same on both sides
Merge joiner suffers from the same problem that was revealed by #44547.

### PR DESCRIPTION
**Pull request by yuzefovich**
**colexec: make hashing of integers of different widths the same**

Previously we had custom hash functions for int16, int32, and int64.
This led to hash joiner not being able to correctly find matching
tuples when the types of the inputs were comparable but different. Now
this is fixed by upcasting all integers to int64 and using that value to
compute the hash.

Note that we don't need to concern ourselves with having types from
different families to hash in the same way, so only integers' hashing
function is changed.

Fixes: #44797.

Release note: None

**colexec: fix merge joiner when types of inputs in eq cols are different**

Merge joiner assumes that the types of inputs in equality columns are
the same. However, this isn't always the case - the types have to be
comparable. This problem is now fixed by performing a CAST operation when
there is a "physical" type mismatch. These "cast" operations are
performed internally by the merge joiner, and the results of the casts
are written into temporary vectors (that are reused) which are used
in-place of original uncasted "key" columns.

This required refactoring of the cast template so that there was
a single `cast` function that takes in "from" and "to" coltypes. We will
no longer generate "type-specific" CAST operators, but `cast` function
performs type switch at the batch level, so the performance hit for cast
operators should be negligible.

This commit also adds a couple of int to int casts (for different
widths).

Fixes: #44798.

Release note (bug fix): Previously, CockroachDB would return an internal
error when merge join operation was performed via the vectorized
execution engine in a case when two sides of the join have comparable
but different types in the equality columns (for example, INT2 on the
left and INT4 on the right). Now this is fixed.

**colexec: add benchmark of castOp**

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
First commit is in a separate PR.

I think I like this approach the most. Thanks @asubiotto for the idea!

**Review by asubiotto — APPROVED** [review]
:lgtm: minus questions

> Reviewed 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-:-M-ooIAqEJ3GmuQ6X9IX:b-g2hho2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/cast_tmpl.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-M-omuvx3REJqO6SSxKU:-M-omuvy0PlGSdr5g-za:bnim1e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8517bce288cab46483c3986b95622cddb568bc0f/pkg/sql/colexec/cast_tmpl.go#L73)):*
> ```Go
> var _ interface{} = execgen.UNSAFEGET
> 
> func cast(fromType, toType coltypes.T, inputVec, outputVec coldata.Vec, n uint16, sel []uint16) {
> ```

Do we have some microbenchmarks we can run to just verify that there is no hit?
___
*[pkg/sql/colexec/execplan.go, line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-M-onCuE62urBQgAIfu5:-M-onCuE62urBQgAIfu6:bz5yu0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f37cfec944e584a2765819d9413a91c122fa201/pkg/sql/colexec/execplan.go#L129)):*
> ```Go
> // createJoinOp distinguishing between the two.
> // Note: the passed in 'result' will be modified accordingly.
> func createJoiner(
> ```

Is the removal of this code just because there wasn't much need for it?

**Review by yuzefovich — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-:-M-pkt-IBxS8gJKUJsYZ:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/cast_tmpl.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-M-omuvx3REJqO6SSxKU:-M-pk1U3CSbAi_AuBhjG:b-49k0pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8517bce288cab46483c3986b95622cddb568bc0f/pkg/sql/colexec/cast_tmpl.go#L73)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Do we have some microbenchmarks we can run to just verify that there is no hit?
</blockquote></details>

We didn't. I added third commit to this PR, and the number are:
```
name                                                    old time/op    new time/op    delta
CastOp/useSel=true/hasNulls=true/int_to_float-16          3.34µs ± 2%    4.02µs ± 4%   ~     (p=0.100 n=3+3)
CastOp/useSel=true/hasNulls=false/int_to_float-16          651ns ± 6%     661ns ± 2%   ~     (p=0.700 n=3+3)
CastOp/useSel=false/hasNulls=true/int_to_float-16         4.42µs ± 2%    6.41µs ± 1%   ~     (p=0.100 n=3+3)
CastOp/useSel=false/hasNulls=false/int_to_float-16        1.12µs ± 0%    1.17µs ± 1%   ~     (p=0.100 n=3+3)
CastOp/useSel=true/hasNulls=true/int_to_decimal-16        34.9µs ± 5%    37.2µs ± 3%   ~     (p=0.200 n=3+3)
CastOp/useSel=true/hasNulls=false/int_to_decimal-16       34.8µs ± 6%    36.1µs ± 4%   ~     (p=0.400 n=3+3)
CastOp/useSel=false/hasNulls=true/int_to_decimal-16       67.8µs ± 1%    72.2µs ± 1%   ~     (p=0.100 n=3+3)
CastOp/useSel=false/hasNulls=false/int_to_decimal-16      71.9µs ± 2%    70.8µs ± 0%   ~     (p=0.200 n=3+3)
CastOp/useSel=true/hasNulls=true/float_to_decimal-16       347µs ± 4%     328µs ± 2%   ~     (p=0.100 n=3+3)
CastOp/useSel=true/hasNulls=false/float_to_decimal-16      364µs ± 3%     368µs ± 6%   ~     (p=1.000 n=3+3)
CastOp/useSel=false/hasNulls=true/float_to_decimal-16      663µs ± 3%     663µs ± 2%   ~     (p=1.000 n=3+3)
CastOp/useSel=false/hasNulls=false/float_to_decimal-16     734µs ± 1%     731µs ± 0%   ~     (p=0.700 n=3+3)

name                                                    old speed      new speed      delta
CastOp/useSel=true/hasNulls=true/int_to_float-16        2.46GB/s ± 2%  2.04GB/s ± 4%   ~     (p=0.100 n=3+3)
CastOp/useSel=true/hasNulls=false/int_to_float-16       12.6GB/s ± 5%  12.4GB/s ± 2%   ~     (p=0.700 n=3+3)
CastOp/useSel=false/hasNulls=true/int_to_float-16       1.85GB/s ± 2%  1.28GB/s ± 1%   ~     (p=0.100 n=3+3)
CastOp/useSel=false/hasNulls=false/int_to_float-16      7.32GB/s ± 0%  7.01GB/s ± 1%   ~     (p=0.100 n=3+3)
CastOp/useSel=true/hasNulls=true/int_to_decimal-16       235MB/s ± 5%   220MB/s ± 3%   ~     (p=0.200 n=3+3)
CastOp/useSel=true/hasNulls=false/int_to_decimal-16      236MB/s ± 6%   227MB/s ± 4%   ~     (p=0.400 n=3+3)
CastOp/useSel=false/hasNulls=true/int_to_decimal-16      121MB/s ± 1%   113MB/s ± 1%   ~     (p=0.100 n=3+3)
CastOp/useSel=false/hasNulls=false/int_to_decimal-16     114MB/s ± 2%   116MB/s ± 0%   ~     (p=0.200 n=3+3)
CastOp/useSel=true/hasNulls=true/float_to_decimal-16    23.7MB/s ± 4%  25.0MB/s ± 2%   ~     (p=0.100 n=3+3)
CastOp/useSel=true/hasNulls=false/float_to_decimal-16   22.5MB/s ± 3%  22.3MB/s ± 6%   ~     (p=1.000 n=3+3)
CastOp/useSel=false/hasNulls=true/float_to_decimal-16   12.4MB/s ± 3%  12.4MB/s ± 2%   ~     (p=1.000 n=3+3)
CastOp/useSel=false/hasNulls=false/float_to_decimal-16  11.2MB/s ± 1%  11.2MB/s ± 0%   ~     (p=0.700 n=3+3)
```
so there is a visible hit when the cast is extremely fast, and no changes when the cast is slow. The fast case doesn't concern me because the bottleneck of any query will be in another part (not in the CAST operation), but the numbers for some of the slow casts seem quite low. We might wanna look into that, but that's a separate issue.
___
*[pkg/sql/colexec/execplan.go, line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-M-onCuE62urBQgAIfu5:-M-pkZKZDU9E49RT8vj9:b2e9kw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f37cfec944e584a2765819d9413a91c122fa201/pkg/sql/colexec/execplan.go#L129)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is the removal of this code just because there wasn't much need for it?
</blockquote></details>

Exactly. Before the logic for handling "post-joiner" projections was here as well as filter-handling, but now that joiners output all of their columns, there is no point in having this `createJoiner` method since it doesn't do much anymore.

**Review by yuzefovich — COMMENTED** [review]
bors r-

I guess I should wait for your approval since I added another commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-:-M-pl4N5CIuTH2rYJyys:bbn9abd)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

**Review by asubiotto — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 4 files at r1, 8 of 8 files at r3, 8 of 8 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44942#-:-M-tsz1WDcN7kcsu8TLN:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

**Comment by yuzefovich** [comment]
TFTR!

bors r+

**Pull request merged**
